tsl/src/nodes/decompress_chunk/exec.c

```diff
@@ -475,6 +477,307 @@ decompress_chunk_begin(CustomScanState *node, EState *estate, int eflags)
+				if (unlikely(pg_mul_s64_overflow(intvalue, amount, &batch_sum)))
+				{
+					ereport(ERROR,
+							(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+							 errmsg("bigint out of range")));
+				}
+
+				/* Add the value to our sum */
+				if (unlikely(pg_add_s64_overflow(result_sum, batch_sum, ((int64 *) &result_sum))))
+					ereport(ERROR,
+							(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+							 errmsg("bigint out of range")));
+			}
+		}
+	}
+	else if (column_description->type == COMPRESSED_COLUMN)
+	{
+		Assert(chunk_state->enable_bulk_decompression);
+		Assert(column_description->bulk_decompression_supported);
+
+		/* Due to the needed manipulation of the target list to emit partials (see
+		 * decompress_chunk_plan_create), PostgreSQL is not able to determine the type of the
+		 * compressed column automatically. So, correct the column type to the correct value. */
+		Assert(list_length(aggref->args) == 1);
+
+#ifdef USE_ASSERT_CHECKING
+		TargetEntry *tlentry = (TargetEntry *) linitial(aggref->args);
+		Assert(IsA(tlentry->expr, Var));
+		Var *input_var = castNode(Var, tlentry->expr);
+		Assert(input_var->vartype == INT4OID);
+#endif
+
+		column_description->typid = INT4OID;
+
+		while (true)
+		{
+			TupleTableSlot *compressed_slot =
+				ExecProcNode(linitial(chunk_state->csstate.custom_ps));
+			if (TupIsNull(compressed_slot))
+			{
+				/* All compressed batches are processed. */
+				break;
+			}
+
+			/* Decompress data */
+			bool isnull;
+			Datum value =
+				slot_getattr(compressed_slot, column_description->compressed_scan_attno, &isnull);
+
+			Ensure(isnull == false, "got unexpected NULL attribute value from compressed batch");
+
+			/* We have at least one value */
+			decompressed_scan_slot->tts_isnull[0] = false;
+
+			CompressedDataHeader *header = (CompressedDataHeader *) PG_DETOAST_DATUM(value);
+			ArrowArray *arrow = NULL;
+
+			DecompressAllFunction decompress_all =
+				tsl_get_decompress_all_function(header->compression_algorithm);
+			Assert(decompress_all != NULL);
+
+			MemoryContext context_before_decompression =
+				MemoryContextSwitchTo(chunk_state->bulk_decompression_context);
+
+			arrow = decompress_all(PointerGetDatum(header),
+								   column_description->typid,
+								   batch_state->per_batch_context);
+
+			Assert(arrow != NULL);
+
+			MemoryContextReset(chunk_state->bulk_decompression_context);
+			MemoryContextSwitchTo(context_before_decompression);
+
+			//#define USE_CUSTOM_SUM_IMPL
+			//#define FORCE_SIMD
+
+#if defined(USE_CUSTOM_SUM_IMPL)
+#define SUM_LANES 4
+
+#if defined(FORCE_SIMD)
+			/* Int wide data type of 16 byte = 128 bit */
+			typedef int v4si __attribute__((vector_size(16)));
+			v4si lane_result = { 0, 0, 0, 0 };
+			v4si lane_buffer = { 0, 0, 0, 0 };
+#else
+			/* Use 32-bit values to be able to fill a 128-bit register with 4 lanes */
+			int32 lane_result[] = { 0, 0, 0, 0 };
+			int32 lane_buffer[] = { 0, 0, 0, 0 };
+#endif
```

Comment:
I remember you said all implementation are similar in performance, let's keep the simplest one for now?

Reply:
The `FORCE_SIMD` implementation is a little faster (since it forces the use of the `xmm0` register and a `paddd` SIMD operation). However, I don't think the small performance gain justifies the complex implementation. So, I would be happy to merge the simplest one.

Reply:
Removed the two alternative implementations from the PR.